# Cours Accéléré de Python
## Configuration de Votre Environnement
Comme vous le savez probablement, configurer votre environnement peut être l'une des parties les plus difficiles lors de l'apprentissage d'un nouvel outil. Heureusement, Python le rend très facile grâce à son gestionnaire de paquets intégré `pip`. D'autres gestionnaires de paquets existent également, tels que `conda`, qui est très souvent utilisé.

Il est recommandé de maintenir **un environnement isolé** pour chaque projet, afin d'éviter les conflits avec l'environnement système de base. Dans la première partie de ce tutoriel, nous verrons comment faire cela pour :
- Google Colab
- un environnement local sur votre poste de travail
- une configuration de base pour un environnement SLURM (par exemple, ARNC / Calcul Canada)

Dans Google Colab, nous n'avons pas besoin de choisir quel gestionnaire de paquets utiliser, car il est préconfiguré avec `pip` et crée automatiquement un nouvel environnement pour chaque session (ce qui est à la fois une bénédiction et une malédiction).

Pour un environnement local sur votre poste de travail, `conda` est recommandé car il s'agit d'un gestionnaire de paquets plus puissant que `pip`, et il permet d'utiliser facilement différentes versions de Python pour différents projets. Conda prend en charge les commandes `conda install` ainsi que `pip install`.

Dans un environnement SLURM, il est généralement recommandé de créer votre environnement avec `virtualenv` plutôt qu'avec `conda`, car il est plus léger. Les paquets peuvent ensuite être installés avec `pip`. ARNC / Calcul Canada fournit d'excellentes instructions pour [configurer un environnement Python sur leur cluster.](https://docs.alliancecan.ca/wiki/Python)

### Configuration dans Google Colab
Heureusement, Google Colab est livré avec la plupart des bibliothèques dont nous avons besoin pour l'apprentissage machine et la science des données préinstallées. Cependant, il arrive que nous ayons besoin d'accéder à des bibliothèques de code ou à des ensembles de données qui ne sont pas préinstallés. Dans ce cas, Colab facilite leur installation, bien que la méthode soit légèrement différente de celle utilisée pour installer des bibliothèques localement.

Google Colab est basé sur les Jupyter Notebooks, qui ont une syntaxe spéciale pour exécuter des commandes système. Nous pouvons exécuter des commandes bash en utilisant la syntaxe `!command`. Par exemple, nous pouvons vérifier la liste des paquets actuellement installés :

In [ ]:
!pip list

Package                            Version
---------------------------------- ------------------
absl-py                            1.4.0
accelerate                         1.3.0
aiohappyeyeballs                   2.4.4
aiohttp                            3.11.12
aiosignal                          1.3.2
alabaster                          1.0.0
albucore                           0.0.23
albumentations                     2.0.3
ale-py                             0.10.1
altair                             5.5.0
annotated-types                    0.7.0
anyio                              3.7.1
argon2-cffi                        23.1.0
argon2-cffi-bindings               21.2.0
array_record                       0.6.0
arviz                              0.20.0
astropy                            7.0.0
astropy-iers-data                  0.2025.2.3.0.32.42
astunparse                         1.6.3
atpublic                           4.1.0
attrs                              25.1.0
audioread            

Nous pouvons également installer et gérer des bibliothèques logicielles en utilisant pip de la même manière. Cependant, il est préférable d'utiliser la commande spéciale `%pip`, car cela permettra de s'assurer que tout fonctionne correctement avec l'environnement Jupyter notebook.

In [ ]:
# Nous pouvons installer PyTorch Geometric qui sera utilisé pendant le tutoriel sur les GNN
%pip install torch_geometric

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.1/63.1 kB 1.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 18.2 MB/s eta 0:00:00


Souvent, nous ne souhaitons pas voir les journaux d'exécution du processus d'installation, car cela encombre simplement nos notebooks. Nous pouvons utiliser la commande "magie" de cellule `%%capture` pour désactiver l'affichage des journaux d'exécution:

In [ ]:
%%capture
%pip install torch_geometric

### Configuration sur votre Poste de Travail Local
Nous allons maintenant voir comment configurer un environnement sur votre poste de travail local. Tout d'abord, nous allons installer `conda`. Si vous n'avez pas encore de version de `conda` installée, suivez les [instructions officielles](https://docs.anaconda.com/miniconda/install/) pour installer Miniconda3. **REMARQUE** Afin d'éviter toute interférence avec l'environnement Colab ou Jupyter, les extraits de code ci-dessous ne sont pas exécutables.

Par exemple, pour une installation sur Linux, nous pouvons copier la commande d'installation rapide fournie par Miniconda :
```bash
mkdir -p ~/miniconda3
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O ~/miniconda3/miniconda.sh
bash ~/miniconda3/miniconda.sh -b -u -p ~/miniconda3
rm ~/miniconda3/miniconda.sh
```

Ensuite, nous pouvons activer l'environnement pour la première fois et le configurer pour qu'il s'active automatiquement lors des futures sessions terminal :
```bash
source ~/miniconda3/bin/activate
conda init --all
```

Ensuite, nous pouvons créer un nouvel environnement pour ce projet :
```bash
conda create -n gnn python=3.11
```
Cela créera un nouvel environnement conda appelé `gnn` utilisant Python 3.11. Nous pouvons ensuite activer l'environnement et installer les bibliothèques nécessaires sans interférer avec d'autres projets.
```bash
conda activate gnn
```

Il est conseillé de garder une trace des bibliothèques nécessaires à votre code en utilisant un fichier `requirements.txt`. Ce fichier liste chaque bibliothèque pip (et facultativement leurs versions) requise sur une ligne distincte. Par exemple, nous pouvons définir un fichier requirements pour le tutoriel GNN comme suit :

```
# requirements.txt
emmet
matplotlib
mp-api
networkx
numpy
pymatgen
scikit-learn
torch
torch_geometric
```

Ensuite, nous pouvons installer ces bibliothèques dans notre environnement conda (après avoir exécuté `conda activate gnn`) avec la commande suivante :
```bash
pip install -r requirements.txt
```

## Notions de Base en Python
### Importation de Modules
L'un des éléments distinctifs de Python est son système de modules. Python facilite grandement le chargement et l'utilisation de bibliothèques logicielles. Certaines bibliothèques sont intégrées, tandis que d'autres doivent être installées auprès de tiers. De plus, il est possible de diviser votre code en plusieurs fichiers et d'"importer" du code depuis un autre fichier.

```
import numpy as np  # ici, nous importons la bibliothèque tierce "numpy" sous l'alias "np"
import torch

# math est une bibliothèque intégrée
from math import sin # nous pouvons importer des fonctions spécifiques en utilisant "from ... import ..."

np.sum([0, 1, 2]), sin(np.pi / 2), torch.rand(1)  # nous pouvons maintenant utiliser les fonctionnalités de numpy, torch ou la fonction "sin" de "math"
```

Python vérifiera d'abord le répertoire actuel pour trouver un fichier correspondant, puis il vérifiera parmi les bibliothèques installées et les modules intégrés.

In [ ]:
import numpy as np  # ici, nous importons la bibliothèque tierce "numpy" sous l'alias "np"
import torch

# math est une bibliothèque intégrée
from math import sin # nous pouvons importer des fonctions spécifiques en utilisant "from ... import ..."

np.sum([0, 1, 2]), sin(np.pi / 2), torch.rand(1)  # nous pouvons maintenant utiliser les fonctionnalités de numpy, torch ou la fonction "sin" de "math"

(3, 1.0, tensor([0.6108]))

Python vérifiera d'abord le répertoire actuel pour trouver un fichier correspondant, puis il vérifiera parmi les bibliothèques installées et les modules intégrés.

### Variables et Types de Données
Comme dans d'autres langages de programmation, Python nous permet de créer et de stocker des données dans des **variables**. Il est important de considérer une variable comme un **conteneur** de données. Les données elles-mêmes peuvent exister même sans la variable, ou être référencées par plusieurs conteneurs. Comprendre ce concept vous aidera à éviter de nombreux bugs courants dans votre code.

En Python, tout est un _objet_ qui peut stocker des données et contenir des méthodes agissant sur ces données. Cela signifie qu'il n'existe **pas de types primitifs** en Python. De plus, Python utilise le concept de **duck typing** ("si ça cancane comme un canard, alors c'est un canard"). Cela signifie que les programmes, en général, ne doivent pas dépendre de types spécifiques, mais plutôt d'objets implémentant des méthodes spécifiques. Nous verrons que de nombreuses fonctions intégrées en Python suivent cette idée.

Nous pouvons créer des variables pour contenir certains des types de données courants en Python comme suit :

Par convention, nous utilisons le format `snake_case` pour les noms de variables en Python.

Remarquez que nous n'avons jamais "déclaré" le type de nos variables. Cela s'explique par le fait que Python n'est pas un langage strictement typé. Nous pouvons même attribuer différents types à une variable au cours du programme. Parfois, cela est utile, mais souvent, cela peut nuire à la lisibilité de nos programmes.

In [ ]:
a = 5  # Nombre, spécifiquement un "int"
b = 3.14  # Nombre, spécifiquement un "float"
c = "hello world"  # "string"
d = 'a'  # Aussi un "string", Python n'a pas de type "char"
e = ["a", "list", "of", "strings"]  # Les listes utilisent des crochets [ ]
f = [1, 3.14, "string"]  # Les listes peuvent mélanger différents types (chaque entrée est juste un "conteneur")
g = True  # Un "bool" prend les valeurs True / False
h = None  # Utilisé de manière similaire à "null" dans d'autres langages
i = { "Drew": 28, "Jama": None}
type(a), type(b), type(c), type(d), type(e), type(f), type(g), type(h), type(i)

(int, float, str, str, list, list, bool, NoneType, dict)

Par convention, nous utilisons le format `snake_case` pour les noms de variables en Python.

Remarquez que nous n'avons jamais "déclaré" le type de nos variables. Cela s'explique par le fait que Python n'est pas un langage strictement typé. Nous pouvons même attribuer différents types à une variable au cours du programme. Parfois, cela est utile, mais souvent, cela peut nuire à la lisibilité de nos programmes.

In [ ]:
a = "yikes"  # maintenant, la variable `a` contient un "string" au lieu d'un entier
type(a)

str

Depuis Python 3.5, il est possible d'inclure des annotations de type pour les variables. Ces annotations sont utilisées pour améliorer la lisibilité et aider l'IDE avec l'autocomplétion, mais elles ne sont pas (en général) appliquées à l'exécution.

In [ ]:
a: int = 5

### Opérations de Base
Python prend en charge les opérations arithmétiques de base sur les nombres.


In [ ]:
a = 1
b = 3
c = a + b  # l'addition d'entiers produit un "int"
c, type(c)

(4, int)

In [ ]:
a = 1.5
b = 4
c = b - a # les opérations avec des "float" produisent un "float"
c, type(c)

(2.5, float)

In [ ]:
a = 7
b = 3
c = a / b  # la division d'entiers produira naturellement un "float"
c, type(c)

(2.3333333333333335, float)

In [ ]:
a = 7
b = 3
c = a // b # nous pouvons effectuer une division entière avec le symbole //
c, type(c)

(2, int)

In [ ]:
a = 6
b = 8
c = a * b  # la multiplication utilise le symbole *
c, type(c)

(48, int)

In [ ]:
a = 2
b = 4
c = a ** b # l'exponentiation utilise le symbole **
c

16

**REMARQUE** Python implémente également le symbole `^`, mais cela effectue un XOR au niveau du bit et non une exponentiation.

In [ ]:
a ^ b

6

L'opérateur modulo / reste est implémenté en utilisant le symbole `%`:


In [ ]:
a = 10
b = 3
c = a % b  # 10 mod 3
c

1

Python nous permet également de définir des nombres en notation scientifique, hexadécimale ou binaire:

In [ ]:
a = 1880  # notation standard
a = 1_880 # nous pouvons séparer les milliers avec un _ pour améliorer la lisibilité
b = 1.88e3  # notation scientifique
c = 0b11101011000 # en binaire
d = 0x758 # en hexadécimal
a == b == c == d

True

### Structure de contrôle et Boucles
Python propose plusieurs structures pour le contrôle de flux, mais peut-être pas autant que certains autres langages.

La structure la plus basique est l'exécution conditionnelle de code avec des instructions `if`, ce qui veut dire "si".

In [ ]:
# Valeur absolue
x = 32
if x >= 0:
  y = x
else:
  y = -x
y

32

In [ ]:
# Valeur absolue
x = -16
if x >= 0:
  y = x
else:
  y = -x
y

16

In [ ]:
animal = 'dog'
if animal == 'cat':
  print('Meow!')
elif animal == 'dog':
  print('Bark!')
elif animal == 'frog':
  print('Ribbit!')
else:
  print('Grrr...')

Bark!


Notez qu'avec la construction `if-elif-else` (si, sinon si, sinon), _exactement_ un bloc est exécuté.

Une autre structure utile pour la structure de contrôle est la boucle. Python propose des boucles du type "foreach" et "while".

In [ ]:
names = ["drew", "jama", "ameer", "mehdi",]
for name in names:  # itérer sur chaque élément de la liste
  print("hello", name.capitalize())

hello Drew
hello Jama
hello Ameer
hello Mehdi


Il n'existe pas de structure comme `for (int i=0; i < 6; i++)` en Python. À la place, la manière idiomatique de faire ce genre de boucle est :

In [ ]:
for i in range(6):
  print(i)

0
1
2
3
4
5


Encore mieux, nous pouvons utiliser la fonction `enumerate` :

In [ ]:
for i, name in enumerate(names):
  print("Hello", name.capitalize(), "you're number", i)

Hello Drew you're number 0
Hello Jama you're number 1
Hello Ameer you're number 2
Hello Mehdi you're number 3


Nous pouvons également exécuter un bloc de code de manière répétée tant qu'une condition est vraie :

In [ ]:
names = ["drew", "jama", "ameer", "mehdi",]
while len(names) > 0:
  print(names.pop(), len(names))  # retirer et retourner le dernier élément de `names`

mehdi 3
ameer 2
jama 1
drew 0


### Définir des Fonctions

Souvent, nous voulons réutiliser du code. Nous pouvons le faire en définissant des fonctions qui acceptent des paramètres et retournent un résultat.

In [ ]:
def square(x):
  return x * x

for i in range(10):
  print(i, '->', square(i))

0 -> 0
1 -> 1
2 -> 4
3 -> 9
4 -> 16
5 -> 25
6 -> 36
7 -> 49
8 -> 64
9 -> 81


Les fonctions peuvent accepter des arguments positionnels ainsi que des arguments nommés :

In [ ]:
def power(x, p=1):
  return x ** p

x = 2
for i in range(4):
  print(power(x, p=i))

1
2
4
8


Il est également possible de définir des fonctions anonymes appelées "fonctions lambda" :

In [ ]:
x = [-3, 4, 12, -2, 8, 0, 1, 7]
sorted(x, key=lambda x: abs(x)) # trier par valeur absolue

[0, 1, -2, -3, 4, 7, 8, 12]

### Définir des Classes
Les classes sont des objets personnalisés. Elles regroupent des données et des méthodes qui opèrent sur ces données. Elles constituent un excellent moyen d'améliorer la lisibilité et la réutilisabilité de votre code. Nous utiliserons des classes personnalisées tout le temps dans PyTorch pour définir de nouveaux modèles :


In [ ]:
from torch import nn

class CustomModel(nn.Module):  # nous "héritons" de Module, ce qui signifie que nous réutilisons ses fonctionnalités

  def __init__(self, in_size, out_size): # la méthode __init__ sert à initialiser une nouvelle instance, comme un constructeur
    # "super" nous permet d'appeler des méthodes de l'implémentation de base, dans ce cas nn.Module
    super().__init__()

    # self est un paramètre spécial qui fait référence à l'instance actuelle, ici
    # nous l'utilisons pour stocker des données personnalisées sur l'instance
    self.in_size = in_size
    self.out_size = out_size

    self.fc = nn.Linear(in_size, out_size)
    self.act = nn.ReLU()

  # "forward" est une méthode attendue par nn.Module
  def forward(self, x):
    # Nous pouvons accéder aux données ou aux méthodes du module avec `self.`
    x = self.fc(x)
    x = self.act(x)

    return x

### Quelques Autres Fonctionnalités
Comme dans la plupart des langages, il est possible que des opérations lèvent des erreurs lorsqu'elles rencontrent des entrées inattendues ou d'autres problèmes. Nous pouvons encapsuler le code susceptible de lever des erreurs dans un bloc `try` et les capturer avec `except`. Cela nous permet d'écrire du code qui gère les erreurs de manière élégante.


In [ ]:
try:
  x = 10 / 0
except ZeroDivisionError:
  print("Oups, impossible de diviser par 0...")
  x = 0
except:
  print("quelque chose s'est mal passé")
  x = -1
x  # la valeur de x est maintenant 0

Opps, can't divide by 0...


0

Python prend en charge nativement l'idée de "gestionnaires de contexte". Cela nous permet d'exécuter un code spécifique au début d'un bloc de code et, à nouveau, lorsque le programme quitte le bloc (peu importe si des erreurs sont levées) :

In [ ]:
with open('/content/sample_data/README.md') as f:  # un certain code est exécuté ici, et le résultat est fourni dans la variable "f"
  txt = f.readline()

# une logique de sortie est exécutée implicitement ici

print(txt)

This directory includes a few sample datasets to get you started.



Dans ce cas, en entrant dans le contexte `with`, nous ouvrons le fichier spécifié et retournons une référence à l'objet fichier dans la variable `f`. Lorsque nous quittons le contexte `with`, `f.close()` sera appelé automatiquement.

Nous verrons souvent ce modèle `with` dans PyTorch comme suit :
```python
with torch.no_grad():  # le calcul des gradients est désactivé, ce qui économise de la mémoire
  model.eval()
  ... # évaluer notre modèle
# le calcul des gradients est maintenant réactivé (s'il était activé avant d'entrer dans le bloc)
```

En interne, le contexte `with` fonctionne en appelant les méthodes `__enter__` et `__exit__` sur l'objet qui lui est transmis. Cela signifie que vous pouvez définir vos propres gestionnaires de contexte personnalisés en implémentant ces méthodes dans votre classe.

### Numpy et PyTorch
Numpy et PyTorch sont deux bibliothèques très utiles en apprentissage automatique.

Numpy implémente des "arrays" qui sont une forme plus efficace et multidimensionnelle de liste. De plus, elle définit de nombreuses opérations mathématiques utiles.

In [ ]:
# Tout d'abord, nous nous assurons que numpy est importé
import numpy as np

In [ ]:
x = np.zeros((6, 6)) # nous pouvons créer une matrice 6x6 de zéros
x

array([[0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.]])

In [ ]:
x.shape, x.dtype # la forme est 6 lignes par 6 colonnes, le type de donnée est un float de 64 bits par défaut

((6, 6), dtype('float64'))

In [ ]:
x = np.ones((3, 8, 8)) # nous ne sommes pas limités aux tableaux à 2 dimensions...
x

array([[[1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.]],

       [[1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.]],

       [[1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1

In [ ]:
x = np.random.rand(3, 4, 4) # crée un tableau 3x4x4 de nombres aléatoires entre [0, 1]
x

array([[[0.45332707, 0.87913782, 0.7525274 , 0.53696492],
        [0.34877662, 0.15144237, 0.579431  , 0.90932915],
        [0.73254163, 0.43978424, 0.98839531, 0.32208989],
        [0.33828115, 0.14894458, 0.19692988, 0.17602081]],

       [[0.16872818, 0.97280371, 0.53619607, 0.77044525],
        [0.33664915, 0.01134832, 0.87576268, 0.76654076],
        [0.61105245, 0.99021261, 0.85370909, 0.55973741],
        [0.9660159 , 0.55608984, 0.29727331, 0.06005138]],

       [[0.09227107, 0.09683557, 0.3737085 , 0.81047916],
        [0.09499326, 0.58141306, 0.28238224, 0.12204229],
        [0.77462464, 0.47719523, 0.23991498, 0.17722404],
        [0.09535425, 0.12413666, 0.95909546, 0.5533829 ]]])

In [ ]:
y = x.sum(axis=0) # somme le long du premier axe (3 éléments)
y.shape, y

((4, 4),
 array([[0.71432631, 1.9487771 , 1.66243197, 2.11788933],
        [0.78041903, 0.74420376, 1.73757591, 1.79791219],
        [2.11821872, 1.90719208, 2.08201938, 1.05905133],
        [1.39965131, 0.82917109, 1.45329865, 0.7894551 ]]))

In [ ]:
x.mean() # retourne la valeur moyenne de tous les éléments, nous nous attendons à ce que cela soit proche de 0.5

0.482116526002723

Numpy (et PyTorch) prennent en charge l'indexation avancée pour les arrays / tenseurs :

In [ ]:
x[:, 0] # retourne le premier élément le long de la 2e dimension

array([[0.46505492, 0.69657101, 0.19713386, 0.93933096],
       [0.90201033, 0.00325078, 0.96537857, 0.40237639],
       [0.69278739, 0.4009709 , 0.1567754 , 0.26029281]])

In [ ]:
x[..., :3]  # retourne uniquement les 3 premiers éléments le long de la dernière dimension

array([[[0.45332707, 0.87913782, 0.7525274 ],
        [0.34877662, 0.15144237, 0.579431  ],
        [0.73254163, 0.43978424, 0.98839531],
        [0.33828115, 0.14894458, 0.19692988]],

       [[0.16872818, 0.97280371, 0.53619607],
        [0.33664915, 0.01134832, 0.87576268],
        [0.61105245, 0.99021261, 0.85370909],
        [0.9660159 , 0.55608984, 0.29727331]],

       [[0.09227107, 0.09683557, 0.3737085 ],
        [0.09499326, 0.58141306, 0.28238224],
        [0.77462464, 0.47719523, 0.23991498],
        [0.09535425, 0.12413666, 0.95909546]]])

In [ ]:
x[[True, False, False]] # Indexation booléenne

array([[[0.46505492, 0.69657101, 0.19713386, 0.93933096],
        [0.39411086, 0.67282509, 0.92814916, 0.0061341 ],
        [0.63186582, 0.43728924, 0.46777357, 0.91862493],
        [0.98333132, 0.38720654, 0.45624926, 0.76001125]]])

In [ ]:
x[:, [1, 2, 0], :]  # Indexation avec des int

array([[[0.39411086, 0.67282509, 0.92814916, 0.0061341 ],
        [0.63186582, 0.43728924, 0.46777357, 0.91862493],
        [0.46505492, 0.69657101, 0.19713386, 0.93933096]],

       [[0.64400218, 0.13739824, 0.39569877, 0.41866399],
        [0.05900367, 0.19063073, 0.89275787, 0.31559409],
        [0.90201033, 0.00325078, 0.96537857, 0.40237639]],

       [[0.49810344, 0.25775015, 0.56243913, 0.11910501],
        [0.73592121, 0.54315779, 0.87011818, 0.61196139],
        [0.69278739, 0.4009709 , 0.1567754 , 0.26029281]]])

PyTorch prend en charge de nombreuses opérations similaires à Numpy avec une interface très proche. Cependant, PyTorch fournit des fonctionnalités supplémentaires très utiles pour l'apprentissage automatique.

Dans PyTorch, ce qui était appelé un "array" dans Numpy est désormais appelé un "tensor". Les tenseurs peuvent exister sur différents appareils, notamment le CPU ou le GPU (cuda). Cela nous permet de tirer parti des capacités de parallélisme extrême des GPU pour nos opérations sur les tenseurs.


In [ ]:
import torch


# REMARQUE : pour utiliser "cuda", nous avons besoin d'un GPU disponible. Cela peut être fait en modifiant le
#            runtime dans Google Colab en allant sur Runtime -> Change Runtime Type
device = "cuda" if torch.cuda.is_available() else "cpu"

x = torch.rand((3, 4, 4))
x

tensor([[[0.5352, 0.0947, 0.7733, 0.0189],
         [0.3561, 0.2553, 0.3326, 0.0998],
         [0.7704, 0.8500, 0.9524, 0.7296],
         [0.7963, 0.3886, 0.5943, 0.7973]],

        [[0.4594, 0.5803, 0.7034, 0.1211],
         [0.4443, 0.5238, 0.9516, 0.2728],
         [0.2328, 0.7155, 0.5388, 0.1192],
         [0.2916, 0.7087, 0.3766, 0.9178]],

        [[0.6944, 0.1613, 0.5245, 0.5382],
         [0.4946, 0.5127, 0.8687, 0.2482],
         [0.2628, 0.6253, 0.0133, 0.6206],
         [0.2462, 0.4264, 0.2256, 0.2895]]])

In [ ]:
x = x.to(device) # nous pouvons déplacer le tenseur sur un autre appareil, si disponible
x

tensor([[[0.5352, 0.0947, 0.7733, 0.0189],
         [0.3561, 0.2553, 0.3326, 0.0998],
         [0.7704, 0.8500, 0.9524, 0.7296],
         [0.7963, 0.3886, 0.5943, 0.7973]],

        [[0.4594, 0.5803, 0.7034, 0.1211],
         [0.4443, 0.5238, 0.9516, 0.2728],
         [0.2328, 0.7155, 0.5388, 0.1192],
         [0.2916, 0.7087, 0.3766, 0.9178]],

        [[0.6944, 0.1613, 0.5245, 0.5382],
         [0.4946, 0.5127, 0.8687, 0.2482],
         [0.2628, 0.6253, 0.0133, 0.6206],
         [0.2462, 0.4264, 0.2256, 0.2895]]], device='cuda:0')

In [ ]:
x.sum(dim=1)  # "axis" est maintenant appelé "dim" dans pytorch

tensor([[2.4579, 1.5886, 2.6526, 1.6456],
        [1.4280, 2.5284, 2.5705, 1.4308],
        [1.6981, 1.7257, 1.6321, 1.6965]], device='cuda:0')

De plus, PyTorch prend en charge la **différentiation automatique** (autograd) pour calculer automatiquement les gradients, et inclut de nombreuses classes et fonctions utiles pour l'apprentissage profond. La fonctionnalité autograd est activée en définissant `requires_grad=True` sur un tenseur. Nous pourrons ensuite appliquer la **rétropropagation du gradient (backpropagation)** pour calculer le gradient d'une fonction par rapport à ce tenseur. Les modules dans `torch.nn` définissent automatiquement cela pour leurs paramètres :


In [ ]:
x = torch.randn((3, 5, 5), requires_grad=True)  # distribution normale moyenne=0, std=1
loss = (x**2).mean()
loss.backward() # calcule les gradients par rapport à "x"
x.grad

tensor([[[ 0.0891,  0.0070, -0.0238, -0.0108,  0.0372],
         [ 0.0246, -0.0149, -0.0119, -0.0057,  0.0094],
         [-0.0005, -0.0100,  0.0006,  0.0565, -0.0198],
         [-0.0279, -0.0247,  0.0073, -0.0178, -0.0113],
         [-0.0294,  0.0098, -0.0086,  0.0331,  0.0269]],

        [[ 0.0407,  0.0506,  0.0296, -0.0104,  0.0140],
         [ 0.0007, -0.0209, -0.0080,  0.0552,  0.0011],
         [-0.0251, -0.0099,  0.0094,  0.0114, -0.0690],
         [ 0.0130,  0.0164,  0.0124, -0.0065, -0.0140],
         [ 0.0186,  0.0462,  0.0169, -0.0002,  0.0350]],

        [[ 0.0341,  0.0105, -0.0002,  0.0191, -0.0138],
         [-0.0165,  0.0001, -0.0308, -0.0374, -0.0460],
         [-0.0775,  0.0375, -0.0406, -0.0025, -0.0117],
         [-0.0118, -0.0352,  0.0542, -0.0090,  0.0196],
         [ 0.0214,  0.0136, -0.0044, -0.0396,  0.0471]]])


Autograd nous permet de construire des architectures de modèles avancées et de les entraîner avec l'algorithme du **descente de gradient** sans nous soucier du calcul manuel des gradients.

Divers optimisateurs (par exemple, `SGD`, `Adam`, etc.) sont inclus dans le module `torch.optim` et peuvent être utilisés pour optimiser une liste de `nn.Parameter` (encapsule un tenseur pour que PyTorch sache que nous voulons l'optimiser) appartenant à un modèle.

Les modules les plus utiles dans PyTorch sont :

- `torch.nn` : Implémente de nombreuses couches de réseaux de neurones et des fonctions utiles. `nn.Module` forme la base de tous les réseaux neuronaux que nous concevons.
- `torch.optim` : Implémente diverses variantes de descente de gradient.
- `torch.utils.data` : Implémente une interface standard pour `Dataset` et fournit `DataLoader` pour itérer dynamiquement sur les données par lots.

Quelques autres bibliothèques utiles incluent :
- `torch_geometric` : Implémente les réseaux neuronaux graphiques. Nous utiliserons cela dans l'atelier GNN !
- `torchvision` : Utile pour les tâches de vision par ordinateur
- `torchaudio` : Utile pour les tâches audio et de traitement du signal
- `matplotlib` / `seaborn` : Bibliothèques communes pour les graphiques
- `pandas` : Travailler avec des données tabulaires
- `dask` : Big data, traitement parallèle sur CPU (avec prise en charge de divers clusters HPC)